### PR TITLE
New script to calculate a global temperature average

### PR DESCRIPTION
**Pull request by burakbayramli**

### CODE REVIEW

**Comment by atrisovic** [comment]
Thanks so much for the PR! @burakbayramli

File: example/code/glob_avg.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+'''
+Creates a monthly temparature time series for a global average
+temperature, using of all Berkeley region data per time index. The
+average is of the temperature absolute value, the anomaly plus the
```
Comment (jrising) [inline example/code/glob_avg.py:4]: Not the absolute value. "The script combines the temperature anomaly with the climatology for each time/region."?

File: example/code/glob_avg.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+base value for a time/region.  Missing data is interpolated using
+Pandas interpolate(). Data is from http://berkeleyearth.org/data/, see
+"Monthly Land" under the Gridded section, and use the "Equal Area"
```
Comment (jrising) [inline example/code/glob_avg.py:7]: So, this is just the average land temperature, right? Should clarify that at the top.

File: example/code/glob_avg.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+avg = pd.DataFrame(df.median(axis=0))
+avg['year'] = avg.apply(lambda x: year(time[x.name]),axis=1)
+avg['mon'] = avg.apply(lambda x: month(time[x.name]),axis=1)
+avg['mon'] = avg['mon'] + 1
+avg['dt'] = avg.apply(lambda x: pd.to_datetime("%d-%02d-01" %(x.year,x.mon)), axis=1)
+avg = avg.set_index('dt')
+avg = avg[0]
+avg = avg[avg.index > '1900-01-01']  
+avg = avg.rolling(50).mean()
```
Comment (jrising) [inline example/code/glob_avg.py:53]: Is this years or months? Either way, 50 is a non-standard averaging period.

File: example/code/glob_avg.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+def year(float_year):
+    year = int(float_year)
+    return year
+
+def month(float_year):
+    month = int((float_year % 1) * 12)
+    return month
+
+monidxs = np.array(list(map(month, time)))
```
Comment (jrising) [inline example/code/glob_avg.py:29]: What do these values look like? I'm concerned about the month calculation, which assumes that months are equally spaced.

File: example/code/glob_avg.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    real_temps = base_temps + anom[:, region]
+    tmp = pd.Series(real_temps)
+    tmp = tmp.replace('--', np.nan)
+    tmp = tmp.interpolate(method='linear',limit_direction='backward')
+    res.append(list(tmp))
+
+arr = np.array(res)
+df = pd.DataFrame(res)
+
+avg = pd.DataFrame(df.median(axis=0))
```
Comment (jrising) [inline example/code/glob_avg.py:45]: Add comments about the structure of the data in `df` and `avg`.

**Review by jrising — CHANGES_REQUESTED** [review]
Thanks for the script! Just a few comments.

File: example/code/glob_avg.py (hunk 2 above)
Comment (burakbayramli) [reply example/code/glob_avg.py:7]: Done

File: example/code/glob_avg.py (hunk 1 above)
Comment (burakbayramli) [reply example/code/glob_avg.py:4]: Removed the comment

File: example/code/glob_avg.py (hunk 4 above)
Comment (burakbayramli) [reply example/code/glob_avg.py:29]: The code worked fine.. month(1922.33) gives 3, month(1922.12) gives 1, etc.

File: example/code/glob_avg.py (hunk 5 above)
Comment (burakbayramli) [reply example/code/glob_avg.py:45]: I'll let the next user print out the contents for that one :)

**Comment by ks905383** [comment]
Hi - thanks for the script!

I just had a few questions on this - 

- What is the goal of the script? Berkeley Earth has a global average time series calculated and easily available from their data page, with presumably their methodologies for gap filling (in line with our current philosophy in the tutorial that we shouldn't encourage end user interpolation if interpolated products already exist). 
- I may be misunderstanding, but why is the script taking a global average of regional averages? (Or, I can't seem to read where it is taking geographic averages - so is this doing regional monthly/yearly averages instead?) Part of the issue is Berkeley Earth's server seems to be down, so I can't seem to download the example file right now and test the code... I'll try again a bit later. 
- Using the base `netCDF4` package is pretty clunky these days - most of this code can be done in a few lines using `xarray` (for example, using their [built-in](http://xarray.pydata.org/en/stable/examples/monthly-means.html) monthly / yearly averaging functionality). This is also more in line with the rest of the tutorial, where we generally show python examples in `xarray` for its readability and functionality. I'm more than happy to provide an example, but like I mentioned above, I can't get into Berkeley Earth's data right now. Once I have that file I can provide some more input. 

Some of these questions would be easily answered if I had access to the `Complete_TAVG_EqualArea.nc` file - does someone happen to have it available and could send it over to me in the meantime while Berkeley's servers get back online?

**Comment by burakbayramli** [comment]
Hi, the only Berkeley Earth global time series I could find was on Kaggle and time series stops after a certain year. Which one wld you be referring to? I needed an up-to-date time series that's why I had to write this script. I contacted Berkeley admin, he didn't know abt my request, but he did give me some advice on why their region data is patchy at places (it was throwing off my average), coverage of Antarctica started pretty late it turns out. 

My script loops through all regions, and for each region it sums base temperature plus anomaly, giving the measured temperature. Then it averages all time series for all regions.

**Comment by ks905383** [comment]
Hi @burakbayramli , apologies for the delay. 

On the same Berkeley Earth data page, there are global mean temperatures under the section "Land + Ocean (1850-Recent)" (e.g. "[Monthly Global Average Temperature (annual summary)](http://berkeleyearth.lbl.gov/auto/Global/Land_and_Ocean_complete.txt)" that provides average global land data until about September 2021 currently (so it looks like it takes 1-2 months to update with the latest values). 

Is there a particular reason why you need the Berkeley Earth data? They limit themselves to land station data (the 'global' product includes data using an ocean model) in their own products. Otherwise any of the data products on the NCAR Climate Data Guide's [page](https://climatedataguide.ucar.edu/climate-data/global-temperature-data-sets-overview-comparison-table) on Global Climate Temperature Datasets may be useful. HadCRUT5 and GISTEMP for example have NetCDFs of global temperature time series, which is what I think is the data you may need? (and they're both a little easier to work with than Berkeley Earth's somewhat clunky data setup). HadCRUT5 is updated until July of this year; GISTEMP until October. 

Like I said, I would counsel against interpolating data if there exist versions of the data that were interpolated by the data provider, at least without good cause. Their methods are generally more sophisticated and informed by their knowledge of their input data and its biases.

**Comment by burakbayramli** [comment]
Hi Kevin, I looked at that file, it has the anomaly values for year-month pairs,  for 1850 January for example the anomaly is  -0.754.. How do I calculate the measured, actual temparature from this? Am I supposed to add the anomaly to the values seen "Jan 1951-Dec 1980 monthly absolute temperature" list in the comments, using the month? 

It is confusing why this sci area always shares data in the form of anomalies. I need an actual temperature measurement, in a simple file, per year per month. But I can do the extra work if I know how to calculate the needed result :)

**Pull request closed without merging**